File: low-level-design/design-patterns/python/behavioural/state-pattern/stopped_state.py
from state import VMState


class StoppedError(Exception):
    pass


class StoppedState(VMState):
    def start(self, vm):
        from state_factory import StateFactory
        print(f"Starting VM {vm.name}")
        vm.set_state(StateFactory.get_starting_state())
    
    def stop(self, vm):
        print(f"VM {vm.name} is already in stopped state, no action performed")
        return StoppedError("VM is already in stopped state, no action performed")
    
    def restart(self, vm):
        from state_factory import StateFactory
        print(f"Restarting VM {vm.name}")
        vm.set_state(StateFactory.get_restarting_state())
    
    def suspend(self, vm):
        print(f"Cannot suspend VM {vm.name} as it is in stopped state")
        return StoppedError("Cannot suspend VM as it is in stopped state")
    
    def get_name(self):
        return "Stopped"

File: low-level-design/design-patterns/golang/behavioural/state-pattern/state/stopped_state.go
package state

import (
	"errors"
	"fmt"
)

func NewStoppedState() *StoppedState {
	return &StoppedState{}
}

type StoppedState struct {
}

func (s *StoppedState) Start(vm *VM) error {
	fmt.Printf(" starting the vm %s \n", vm.name)
	vm.SetState(NewStartingState())
	return nil
}
func (s *StoppedState) Stop(vm *VM) error {
	fmt.Printf(" %s vm is already stopped \n", vm.name)
	return errors.New("vm is already stopped")
}
func (s *StoppedState) Restart(vm *VM) error {
	fmt.Printf(" restarting the vm %s \n", vm.name)
	vm.SetState(NewStartingState())
	return nil
}
func (s *StoppedState) Suspend(vm *VM) error {
	fmt.Printf(" %s cannot suspend the vm as it is already stopped \n", vm.name)
	return errors.New("cannot suspend the vm as it is already stopped")
}

func (s *StoppedState) GetName() string {
	return "Stopped"
}
